SignalServiceKit/src/Messages/OWSMessageSender.m

```diff
@@ -988,7 +988,16 @@ - (void)sendMessage:(TSOutgoingMessage *)message
             // of new secondary devices when this message send fails.
             DDLogWarn(@"%@ sync message has no device messages but account has secondary devices.", self.tag);
         } else if (hasDeviceMessages) {
-            OWSFail(@"%@ sync message has device messages for unknown secondary devices.", self.tag);
+            for (NSDictionary *deviceMessage in deviceMessages) {
+                NSNumber *_Nullable deviceId = deviceMessage[@"destinationDeviceId"];
+                if (deviceId && [deviceId isKindOfClass:[NSNumber class]] && [deviceId isEqualToNumber:@(1)]) {
```

Comment:
>  trying to sending sync message to primary device, presumably because user has just registered.

I don't understand why a device would ever send a message to itself, *just registered* or otherwise. Can you expound?

Reply:
Because one device messages is created for each `devices` in `SignalRecipient`.  `SignalRecipient`'s device list defaults to a single entry for the primary device `1`, even if the  `SignalRecipient` represents yourself.  When a newly-registered user sends their first sync message (from their primary), it'll just be a single device message - to their own primary.  They'll get a `409` response code and update the actual device list for themselves (presumably empty) at that time.   

We could special-case the default `devices` list for the `SignalRecipient` that represents ourself, but I was looking for the safest, cheapest change in this critical code path.

Reply:
OK I now understand how this fixes an overzealous assert upon registration.

> We could special-case the default devices list for the SignalRecipient that represents ourself, but I was looking for the safest, cheapest change in this critical code path.

FWIW this proposal seems easier to understand. We never want to send a message to our own device.

If I understood our conversation earlier, you said this PR also fixes linking new devices - before this change linked devices wouldn't receive new messages (in master). I don't understand how this fixes that.

And stepping back a bit, if we're already involving OWSDevice, why are we using Recipient.devices at all? Why isn't OWSDevices on it's own enough to determine if we should be sending sync messages?

Though I have some concurrency concerns around the way we're doing that in master right now, see https://github.com/WhisperSystems/Signal-iOS/pull/2537#discussion_r149211898, it seems more straight forward to me to fix that, vs. the current approach in master which uses some combination of OWSDevices and Recipient.devices.

Reply:
Another argument for building towards something like `DeviceManager.shared.hasLinkedDevice` is that we can more easily continue to save resources.

e.g. the current solution prevents sending the message itself, but it doesn't prevent (e.g.) uploading the attachment that we're inevitably *not* going to sync

Reply:
Hm... looking further I'm having second thoughts about my proposal...  We actually don't know that our local list of OWSDevice is complete, e.g. if we start the linking process, then kill the iOS app, then complete linking on desktop, iOS won't know about that device unless we re-visit the devices screen. We could do more work to make sure we're up to date, but at that point, it's not as simple as I was advocating.

Maybe that's why you're proposing inspecting Recipient.devices, which may not be current, but will be *corrected* by the server as we send messages...

Reply:
> FWIW the "special-case" the SignalRecipient.selfRecipient proposal seems easier to understand. We never want to send a message to our own device.

Okay, I've taken that approach - and greatly elaborated the comments here.

Reply:
> If I understood our conversation earlier, you said this PR also fixes linking new devices - before this change, linked devices wouldn't receive new messages (in master). I don't understand how this fixes that.

Check out the new comments - is it clear now?

Reply:
> Maybe that's why you're proposing inspecting Recipient.devices, which may not be current, but will be corrected by the server as we send messages...

That's exactly right - for the purposes of sending, we want the service to be the single source of truth.  Not least, that means we could someday link devices from - for example - a linked iPad and the primary device would learn of that linked device the next time it tried to send a sync message.